core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,157 @@
+import io.trino.sql.tree.Cast;
+import io.trino.sql.tree.CoalesceExpression;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.FunctionCall;
+import io.trino.sql.tree.GenericLiteral;
+import io.trino.sql.tree.Node;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.trino.sql.planner.plan.Patterns.mergeProcessor;
+import static io.trino.sql.planner.plan.Patterns.mergeWriter;
+import static io.trino.sql.planner.plan.Patterns.project;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static io.trino.sql.planner.plan.Patterns.tableFinish;
+import static io.trino.sql.planner.plan.Patterns.tableScan;
+import static java.util.Objects.requireNonNull;
+
+public class PushMergeWriterUpdateIntoConnector
+        implements Rule<TableFinishNode>
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+    private static final Capture<ProjectNode> PROJECT_NODE_CAPTURE = newCapture();
+    private static final Capture<MergeWriterNode> MERGE_WRITER_NODE_CAPTURE = newCapture();
+
+    private static final Pattern<TableFinishNode> PATTERN =
+            tableFinish().with(source().matching(
+                    mergeWriter().capturedAs(MERGE_WRITER_NODE_CAPTURE).with(source().matching(
+                            mergeProcessor().with(source().matching(
+                                    project().with(source().matching(
+                                            tableScan().capturedAs(TABLE_SCAN))).capturedAs(PROJECT_NODE_CAPTURE)))))));
+
+    public PushMergeWriterUpdateIntoConnector(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer, Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    private final Metadata metadata;
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<TableFinishNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TableFinishNode node, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+        ProjectNode project = captures.get(PROJECT_NODE_CAPTURE);
+        MergeWriterNode mergeWriter = captures.get(MERGE_WRITER_NODE_CAPTURE);
+
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(new Symbol("merge_row")).getChildren();
+
+        if (mergeAssignments.size() < orderedColumnNames.size()) {
```

Comment:
I don't understand this condition. Can you elaborate a little?

Reply:
Sure, here we got merge_row. It usually looks like this - 

it's kind of row from table with 3 additional fields at the end, plus it's ordered and on the places, of base columns could be 2 options - 
like constant (for example 100), it means that we want update of this column with value 100, 
and second option - it's symbol column name, it means, that we don't touch the value.

But there is one case, which I described in tests, when we update all columns in row, in this situation optimiser encode all values to base64. In this case we have instead of (number of columns + 3 ) merge assignments - only 1 merge assignment, I still don't know how to overcome this case, so I described it in Test.
So for now we didn't support UPDATE ALL.

Thats why this check is here, I think it could be replaces with something more meaningful - like check that we have base64 node.

Reply:
replaced with CALL (base64) check

Reply:
https://github.com/trinodb/trino/pull/16445#discussion_r1169712863